radix-engine/src/engine/track.rs

```diff
@@ -446,49 +446,67 @@ impl<'s, R: FeeReserve> Track<'s, R> {
+
+        match result {
+            Ok(()) => Ok(self),
+            Err(error) => Err(PreExecutionError {
+                fee_summary: self.fee_reserve.finalize(),
+                error,
+            }),
+        }
+    }
+
+    fn attempt_apply_pre_execution_costs(
+        &mut self,
+        transaction: &Executable,
+    ) -> Result<(), FeeReserveError> {
+        let encoded_instruction_length =
+            Self::to_u32_or_pre_error(scrypto_encode(transaction.instructions()).len())?;
+        let initial_proofs_count =
+            Self::to_u32_or_pre_error(transaction.auth_zone_params().initial_proofs.len())?;
+        let blobs_length = Self::to_u32_or_pre_error(
+            transaction
+                .blobs()
+                .iter()
+                .map(|b| b.0 .0.len() + b.1.len())
+                .sum::<usize>(),
+        )?;
+
+        self.fee_reserve
             .consume(self.fee_table.tx_base_fee(), "base_fee", true)
             .and_then(|()| {
                 self.fee_reserve.consume(
-                    self.fee_table.tx_manifest_decoding_per_byte()
-                        * transaction.manifest_instructions_size() as u32,
+                    self.fee_table.tx_manifest_decoding_per_byte() * encoded_instruction_length,
                     "decode_manifest",
                     true,
                 )
             })
             .and_then(|()| {
                 self.fee_reserve.consume(
-                    self.fee_table.tx_manifest_verification_per_byte()
-                        * transaction.manifest_instructions_size() as u32,
+                    self.fee_table.tx_manifest_verification_per_byte() * encoded_instruction_length,
                     "verify_manifest",
                     true,
                 )
             })
             .and_then(|()| {
                 self.fee_reserve.consume(
-                    self.fee_table.tx_signature_verification_per_sig()
-                        * transaction.auth_zone_params().initial_proofs.len() as u32,
+                    self.fee_table.tx_signature_verification_per_sig() * initial_proofs_count,
                     "verify_signatures",
                     true,
                 )
             })
             .and_then(|()| {
                 self.fee_reserve.consume(
-                    transaction.blobs().iter().map(|b| b.len()).sum::<usize>() as u32
-                        * self.fee_table.tx_blob_price_per_byte(),
+                    self.fee_table.tx_blob_price_per_byte() * blobs_length,
                     "blobs",
                     true,
                 )
-            });
+            })
+    }
 
-        match result {
-            Ok(()) => Ok(self),
-            Err(error) => Err(PreExecutionError {
-                fee_summary: self.fee_reserve.finalize(),
-                error,
-            }),
-        }
+    fn to_u32_or_pre_error(size: usize) -> Result<u32, FeeReserveError> {
```

Comment:
Very good point... I looked into this more and there are quite a lot of places where we can overflow in the fee reserve and fee table right now. I've added a "remove all overflow risks" task as part of the Recosting epic on the work tracker.

For now, I've introduced `consume_multiplied` and `consume_sized` methods onto the fee reserve, which appear to resolve your concerns quite nicely. But input welcome :)

Reply:
I thought only pre-execution cost skip overflow checks, because we're making assumption about transaction payload size. Where else did you see overflow?